File: frontend/lib/api/practice.ts
import { apiClient } from './client';

export interface PracticeSet {
  id: string;
  title: string;
  domain: string;
  difficulty: string;
  estimated_minutes: number;
}

export interface PracticeQuestion {
  id: string;
  set_id: string;
  title: string;
  problem_statement: string;
  difficulty: string;
  options: any[];
  correct_option: string;
  explanation: string;
}

export interface PracticeAttempt {
  id: string;
  title: string;
  score: number;
  total_questions: number;
  completed_at: string;
  duration_seconds: number;
  domain: string;
}

export const practiceApi = {
  getPracticeSets: async (): Promise<{ sets: PracticeSet[] }> => {
    return apiClient.get('/practice/sets');
  },

  getPracticeSet: async (id: string): Promise<{ set: PracticeSet; questions: PracticeQuestion[] }> => {
    return apiClient.get(`/practice/sets/${id}`);
  },

  createAttempt: async (setId: string, score: number, totalQuestions: number, durationSeconds: number): Promise<{ status: string; attempt_id: string }> => {
    return apiClient.post('/practice/attempts', {
      set_id: setId,
      score,
      total_questions: totalQuestions,
      duration_seconds: durationSeconds
    });
  },

  getLedger: async (): Promise<{ ledger: PracticeAttempt[] }> => {
    return apiClient.get('/practice/ledger');
  }
};

File: backend/app/models/practice.py
from datetime import datetime
from sqlalchemy import Column, String, Integer, DateTime, ForeignKey, JSON
from sqlalchemy.orm import relationship
import uuid
from app.core.database import Base

def generate_uuid():
    return str(uuid.uuid4())

class PracticeSet(Base):
    __tablename__ = "practice_sets"

    id = Column(String, primary_key=True)
    title = Column(String, nullable=False)
    domain = Column(String, nullable=False)
    difficulty = Column(String, nullable=False)
    estimated_minutes = Column(Integer, default=20)


class PracticeQuestion(Base):
    __tablename__ = "practice_questions"

    id = Column(String, primary_key=True)
    set_id = Column(String, ForeignKey("practice_sets.id", ondelete="CASCADE"))
    title = Column(String, nullable=False)
    problem_statement = Column(String, nullable=False)
    difficulty = Column(String, nullable=False)
    options = Column(JSON)
    correct_option = Column(String)
    explanation = Column(String)


class UserPracticeAttempt(Base):
    __tablename__ = "user_practice_attempts"

    id = Column(String, primary_key=True, default=generate_uuid)
    user_id = Column(String, ForeignKey("profiles.id", ondelete="CASCADE"))
    set_id = Column(String, ForeignKey("practice_sets.id", ondelete="CASCADE"))
    score = Column(Integer, default=0)
    total_questions = Column(Integer, default=0)
    duration_seconds = Column(Integer, default=0)
    completed_at = Column(DateTime, default=datetime.utcnow)
